lib/livebook/notebook/explore/kino/pong.livemd

```diff
@@ -21,355 +19,721 @@ Mix.install([
-      #{Enum.join(object_svgs)}
-    </svg>
+    <rect x="#{x * scale}"
+          y="#{y * scale}"
+          width="#{width * scale}"
+          height="#{height * scale}"
+          fill="black" />
     """
   end
 end
 ```
 
-With this piece of code, we can now declaratively compose an image
-out of basic geometric shapes.
-
-At this point we can already experiment and render our static scene!
+And animate it over the `y` dimension, with the addition
+that the paddle must never leave the scene:
 
 ```elixir
-Canvas.new(400, 200)
-|> Canvas.add_rect(0, 0, 400, 200, fill: "#d4f1f477")
-|> Canvas.add_text(200, 4, "0 : 0", font_size: 10)
-|> Canvas.add_rect(0, 10, 2, 40, fill: "black")
-|> Canvas.add_rect(398, 120, 2, 40, fill: "black")
-|> Canvas.add_circle(200, 100, 8, fill: "tomato")
-|> Canvas.to_svg()
-|> Kino.Image.new(:svg)
+defimpl Animation, for: Game.Paddle do
+  def step(paddle, scene) do
+    %{paddle | y: clip(paddle.y + paddle.dy, 0, scene.height - paddle.height)}
+  end
+
+  defp clip(value, min, max), do: value |> max(min) |> min(max)
+end
 ```
 
-## Game logic
+We are ready to give it a try!
+
+<!-- livebook:{"branch_parent_index":4} -->
 
-Having the image in mind, we can now model the game state together
-with all its rules. It makes sense to keep this part separate from
-the runtime elements to make it predictable and easy to test.
-Consequently we will encapsulate the whole game state in a struct
-and define functions for the possible interactions.
+## Mixing control streams
+
+To animate the paddle, we will need two control streams:
+the keyboard control and an interval to refresh
+the paddle as we keep holding the arrow keys up
+and down.
 
 ```elixir
-defmodule Pong do
-  @moduledoc """
-  Represents the paddles game state and rules.
-  """
+# The frame we will render scene images on top of
+frame = Kino.Frame.new() |> Kino.render()
+
+# The scene with a static background
+scene =
+  Scene.new(200, 50)
+  |> Scene.add(Scene.Rect.new(0, 0, 200, 50, fill: "#d4f1f477"))
+
+# The paddle which we will animate
+paddle = Game.Paddle.new(0, 21, 2, 8)
+
+# The keyboard control
+keyboard = Kino.Control.keyboard([:status, :keyup, :keydown]) |> Kino.render()
+
+# For each keyboard every and on every 33ms,
+# emit an event while the keyboard is enabled.
+[keyboard, Kino.Control.interval(33)]
+|> Kino.Control.stream()
+|> Stream.take_while(fn
+  event when event.enabled == false -> false
+  _event -> true
+end)
+|> Enum.reduce(paddle, fn event, paddle ->
+  # Build the image
+  image = scene |> Scene.add(paddle) |> Scene.to_svg() |> Kino.Image.new(:svg)
 
-  defstruct [:state, :left, :right, :ball]
+  # Update the frame
+  Kino.Frame.render(frame, image)
 
-  @type t :: %__MODULE__{state: state(), left: paddle(), right: paddle(), ball: ball()}
+  # Animate the ball
+  paddle
+  |> Game.Paddle.on_key(event)
+  |> Animation.step(scene)
```

Comment:
Shouldn't we do animation step every 33ms rather than on every event? Currently any key event would "fast forward" in away. I'd split it into two clauses, for interval event we step and render, otherwise only update the paddle.

Reply:
Ideally yes, but this is only for checking it works, so I think it is fine for simplicity.